src/op.cpp

```diff
@@ -1,64 +1,217 @@
+    constexpr size_t Size = sizeof...(Args);
+    uint32_t dep[Size] = { args... };
+    Variable *v[Size];
+
+    bool placeholder = false,
+         simplify = false,
+         literal = true;
+
+    JitBackend backend = JitBackend::Invalid;
+    VarType type = VarType::Void;
+    uint32_t size = 0;
+    const char *err = nullptr;
+
+    for (uint32_t i = 0; i < Size; ++i) {
+        if (!dep[i])
+            continue;
+        Variable *vi = jitc_var(dep[i]);
+
+#if !defined(NDEBUG)
+        if constexpr (Flags & IsArithmetic) {
+            if (unlikely(!jitc_is_arithmetic(vi))) {
+                err = "expected arithmetic operand types";
+                goto fail;
+            }
+        }
+
+        if constexpr (Flags & IsInt) {
+            if (unlikely(!jitc_is_int(vi))) {
+                err = "expected integer operand types";
+                goto fail;
+            }
+        }
+
+        if constexpr (Flags & IsIntOrBool) {
+            if (unlikely(!jitc_is_int(vi) && !jitc_is_bool(vi))) {
+                err = "expected integer or boolean operand types";
+                goto fail;
+            }
+        }
+
+        if constexpr (Flags & IsNotVoid) {
+            if (unlikely(jitc_is_void(vi))) {
+                err = "operand cannot be void";
+                goto fail;
+            }
+        }
+
+        if constexpr (Flags & IsFloat) {
+            if (unlikely(!jitc_is_float(vi))) {
+                err = "expected floating point operand types";
+                goto fail;
+            }
+        }
+
+        if constexpr (Flags & IsCUDA) {
+            if (unlikely((JitBackend) vi->backend != JitBackend::CUDA)) {
+                err = "operation is only supported on the CUDA backend";
+                goto fail;
+            }
+        }
+
+        if constexpr (Flags != Disabled) {
+            if (unlikely(type != VarType::Void && (VarType) vi->type != type)) {
+                err = "operands have incompatible types";
+                goto fail;
+            }
+        }
+
+        if (unlikely(backend != JitBackend::Invalid && (JitBackend) vi->backend != backend)) {
+            err = "operands have different backends";
+            goto fail;
+        }
+
+#endif
+
+        size = std::max(size, vi->size);
+        placeholder |= (bool) vi->placeholder;
+        bool is_literal = vi->is_literal();
+        literal &= is_literal;
+        simplify |= is_literal;
+        backend = (JitBackend) vi->backend;
+        if (type == VarType::Void)
+            type = (VarType) vi->type;
+        v[i] = vi;
+    }
+
+    if (size > 0) {
+        // Try simplifying binary expressions with matched arguments
+        if constexpr (Size == 2)
+            simplify |= dep[0] == dep[1];
+
+        for (uint32_t i = 0; i < Size; ++i) {
+            if (unlikely(v[i]->size != size && v[i]->size != 1)) {
+                err = "operands have incompatible sizes";
+                size = (uint32_t) -1;
+                goto fail;
```

Comment:
This might actually be the first time I see a `goto` in real code :smile:

Reply:
I think one is entitled to use one every now and then :D (very handy for a centralized failure/cleanup part of a function). It's very often used in the CPython codebase, e.g.: https://github.com/python/cpython/blob/main/Objects/longobject.c#L1038

Reply:
This rings a bell, for a completely different reason

There was a CppCon talk about some `goto` trickery that could boost performance significantly. I remember the speaker mentionning that CPython made use of this.